src/heuristic/HeuristicMapper.cpp

```diff
@@ -597,3 +615,162 @@ void HeuristicMapper::lookahead(const std::size_t      layer,
+  }
+
+  if (arch.bidirectional()) {
+    costFixed += COST_BIDIRECTIONAL_SWAP;
+  } else {
+    costFixed += COST_UNIDIRECTIONAL_SWAP;
+  }
+}
+
+void HeuristicMapper::Node::applyTeleportation(const Edge&   swap,
+                                               Architecture& arch) {
+  nswaps++;
+  swaps.emplace_back();
+  const auto q1 = qubits.at(swap.first);
+  const auto q2 = qubits.at(swap.second);
+
+  qubits.at(swap.first)  = q2;
+  qubits.at(swap.second) = q1;
+
+  if (q1 != -1) {
+    locations.at(static_cast<std::size_t>(q1)) =
+        static_cast<std::int16_t>(swap.second);
+  }
+  if (q2 != -1) {
+    locations.at(static_cast<std::size_t>(q2)) =
+        static_cast<std::int16_t>(swap.first);
+  }
+
+  std::uint16_t middleAnc = std::numeric_limits<decltype(middleAnc)>::max();
+  for (const auto& qpair : arch.getTeleportationQubits()) {
+    if (swap.first == qpair.first || swap.second == qpair.first) {
+      middleAnc = static_cast<std::uint16_t>(qpair.second);
+    } else if (swap.first == qpair.second || swap.second == qpair.second) {
+      middleAnc = static_cast<std::uint16_t>(qpair.first);
+    }
+  }
+
+  if (middleAnc == std::numeric_limits<decltype(middleAnc)>::max()) {
+    throw QMAPException("Teleportation between seemingly wrong qubits: " +
+                        std::to_string(swap.first) + " <--> " +
+                        std::to_string(swap.second));
+  }
+
+  std::uint16_t source = std::numeric_limits<decltype(source)>::max();
+  std::uint16_t target = std::numeric_limits<decltype(target)>::max();
+  if (arch.getCouplingMap().find({swap.first, middleAnc}) !=
+          arch.getCouplingMap().end() ||
+      arch.getCouplingMap().find({middleAnc, swap.first}) !=
+          arch.getCouplingMap().end()) {
+    source = swap.first;
+    target = swap.second;
+  } else {
+    source = swap.second;
+    target = swap.first;
+  }
+
+  if (source == middleAnc || target == middleAnc) {
+    std::clog << "FAIL: TELE " << source << " -(" << middleAnc << ")-> "
+              << target << "\n";
+    throw QMAPException("Overlap between source/target and middle "
+                        "ancillary in teleportation.");
+  }
+
+  swaps.back().emplace_back(source, target, middleAnc, qc::Teleportation);
+
+  costFixed += COST_TELEPORTATION;
+}
+
+void HeuristicMapper::Node::recalculateFixedCost(const Architecture& arch) {
+  costFixed = 0;
+  for (auto& swapNode : swaps) {
+    for (auto& swap : swapNode) {
+      if (swap.op == qc::SWAP) {
+        if (arch.bidirectional()) {
+          costFixed += COST_BIDIRECTIONAL_SWAP;
+        } else {
+          costFixed += COST_UNIDIRECTIONAL_SWAP;
+        }
+      } else if (swap.op == qc::Teleportation) {
+        costFixed += COST_TELEPORTATION;
+      }
+    }
+  }
+}
+
+void HeuristicMapper::Node::updateHeuristicCost(
+    const Architecture&         arch,
+    const TwoQubitMultiplicity& twoQubitGateMultiplicity,
+    bool                        admissibleHeuristic) {
```

Comment:
Small nit: can be `const` here.

Reply:
You mean `admissibleHeuristic`? Right, I made it `const` now

Reply:
Jap, exactly.